homeassistant/components/russound_rio/config_flow.py

```diff
@@ -0,0 +1,127 @@
+"""Config flow to configure russound_rio component."""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+from typing import Any
+
+from aiorussound import Russound
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PORT
+from homeassistant.helpers import config_validation as cv
+
+from .const import DOMAIN, RUSSOUND_RIO_EXCEPTIONS, NoPrimaryControllerException
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_NAME): cv.string,
+        vol.Optional(CONF_PORT, default=9621): cv.port,
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def find_primary_controller_metadata(controllers):
+    """Find the mac address of the primary Russound controller."""
+    for controller_id, mac_address, controller_type in controllers:
+        # The integration only cares about the primary controller linked by IP and not any downstream controllers
+        if controller_id == 1:
+            return (mac_address, controller_type)
+    return None
+
+
+class FlowHandler(ConfigFlow, domain=DOMAIN):
+    """Russound RIO configuration flow."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by the user."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            name = user_input[CONF_NAME]
+            port = user_input[CONF_PORT]
+
+            try:
```

Comment:
Let's try to keep everything out of the `try` block that can't raise. If I am correct you can connect, enumerate controllers and close and move everything else to the `else:`

Reply:
Moved the set unique ID out of the method. You are correct on those three methods. I also made the find_primary_controller_metadata method return the exception directly so that is able to remain put.